### PR TITLE
Add heartbeat to websocket connection

### PR DESCRIPTION
**Pull request by rhuairahrighairidh**
Adds a heartbeat to the websocket connection to keep the connection alive.

### CODE REVIEW

File: src/ws-reconnect.ts [hunk 1]
```diff
@@ -104,6 +109,10 @@ export class WebSocketReconnector extends EventEmitter2 {
    * `close ()` would not trigger a reconnect.
    */
   close () {
+    if (this._heartbeatTimer) {
+      global.clearTimeout(this._heartbeatTimer)
```
Comment (sentientwaffle) [inline src/ws-reconnect.ts]: `global.` is unnecessary, and I think it will break in the browser.

File: src/ws-reconnect.ts [hunk 2]
```diff
@@ -114,12 +123,26 @@ export class WebSocketReconnector extends EventEmitter2 {
...
   }
 
+  heartbeat () {
+    if (this._heartbeatTimer) {
+      global.clearTimeout(this._heartbeatTimer)
+    }
+
+    if (this._instance && this._instance.readyState === this._instance.OPEN) {
+      this._instance.ping()
+      this._heartbeatTimer = setTimeout(() => this.heartbeat(), 5000)
```
Comment (sentientwaffle) [inline src/ws-reconnect.ts]: `5000` is pretty frequent just to keep a connection alive, isn't it? I'd think 30 seconds would be fine.

File: package.json [hunk 3]
```diff
@@ -14,7 +14,7 @@
   "scripts": {
     "lint": "tslint --project .",
     "build": "tsc --project .",
-    "test": "mocha ./test/index.test.js",
+    "test": "mocha ./test/index.test.js --exit",
```
Comment (sentientwaffle) [inline package.json]: If this change was needed to make the tests exit, I think it indicates that the timers aren't being cleaned up properly -- which is a bug.

**Review by sentientwaffle — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (kincaidoneil) [reply package.json]: Fixed by calling `disconnect` on the plugin in a couple of the tests to remove the timer.

File: src/ws-reconnect.ts (hunk 2 above)
Comment (kincaidoneil) [reply src/ws-reconnect.ts]: Is there somewhere in the middle (e.g. 15-20s)? We experience connection drops every 30 seconds or so on the low end, maybe every 45 seconds on the high end, if the connection is idle.

Comment (sentientwaffle) [reply src/ws-reconnect.ts]: 20s sounds ok. Are you using a google load balancer like ingress? By default they close websockets after 30 seconds (https://cloud.google.com/load-balancing/docs/https/#websocket_proxy_support), which heartbeats won't change.

**Comment by spearl** [comment]
What websocket service are you using that does keep-alive based on websocket activity?

If you're using say a GCP load balancer, the timeout for websockets is the same as the HTTP timeout which is tunable but defaults to 30s. And as @sentientwaffle said above a heartbeat does not change this behavior as it does not inspect websocket activity.

**Comment by kincaidoneil** [comment]
@spearl @sentientwaffle Right now we're just using nginx with no load balancer, so it would help us.

As an aside, Heroku and other providers I think do inspect WebSocket activity:

[https://devcenter.heroku.com/articles/websockets#timeouts](https://devcenter.heroku.com/articles/websockets#timeouts)
> The normal Heroku HTTP routing timeout rules apply to WebSocket connections. Either client or server can prevent the connection from idling by sending an occasional ping packet over the connection.
>
> WebSocket capable endpoints support HTTP keep-alive. Refer to the documentation for the Heroku router for complete details.

[https://devcenter.heroku.com/articles/http-routing#timeouts](https://devcenter.heroku.com/articles/http-routing#timeouts)
> After the initial response, each byte sent (either from the client or from your app process) resets a rolling 55 second window. If no data is sent during this 55 second window then the connection is terminated and a H15 or H28 error is logged.

They also offer [session affinity](https://blog.heroku.com/session-affinity-ga) using HTTP cookies so each the websocket is sticky to the same instance.

**Comment by kincaidoneil** [comment]
Also, if the client's network goes down, but the plugin isn't sending any messages, sometimes it doesn't actually know the connection dropped. We've been encountering this issue with Switch.

A plugin receiving packets cannot be reached, which causes all packets sent to it to be rejected (since the plugin itself wasn't sending any messages, no reconnect would be triggered).

I just tested that this PR resolves that issue, terminating the connection when the ping fails, which triggers a reconnect.

@sentientwaffle Can we merge this?

**Review by sentientwaffle — APPROVED** [review]
Approved, though consider upping the heartbeat interval from 5000 to 20000.

**Comment by kincaidoneil** [comment]
Thanks for the ack, and sorry for the oversight on my part; upped the interval to 20 seconds.

**Pull request merged**